### DISCUSSION TITLE
building with Nuitka

### DISCUSSION THREAD

**Discussion by nschloe** [pypa/cibuildwheel#939] [Q&A] [answered]
I manually build my manylinux wheels with [Nuitka](https://github.com/Nuitka/Nuitka/) as follows
```sh
/opt/python/cp39-cp39/bin/pip install nuitka
/opt/python/cp39-cp39/bin/python setup.py bdist_nuitka
auditwheel repair dist/*.whl
```
Is is possible to use cibuildwheel to use `bdist_nuitka` instead of `bdist`, too?

**Comment by henryiii** [comment, marked as answer]
Custom commands are not supported by upstream anymore (for several years), see https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html. I think you should bring this up with nuitka; they should integrate with `bdist_wheel`, not custom commands. cibulodwheel never runs directly on the setup.py, but uses PEP 517 builders like pypa/pip and pypa/build.

Note that if `pip instal <package>` cannot build from SDist, this is really bad for your users, too; if they don't get a wheel, there should at least be an attempt to build from source.

I think you can fixed this from your end too by replacing the bdist_wheel command with nuitka's bdist_nuitka command in your setup.py.

**Reply by henryiii** [reply]
Ahh, actually, `bdist_nuitka` is just something they use in testing to compare "traditional" wheels with nuitka wheels. They actually set them equal normally:  https://github.com/Nuitka/Nuitka/blob/1bf3c56522bb6aef3c674bcefb8d8bbd43743f6b/nuitka/distutils/DistutilCommands.py#L44

If you do want to support nuitka wheels as well as a normal wheel fallback, I'd just use an environment variable to toggle which bdist command is used as `bdist_wheel` in your setup.py.

**Reply by nschloe** [reply]
How do you override bdist_wheel in setup.py or -- even better -- pyproject.toml?

**Reply by henryiii** [reply]
Setuptools does not have support for pyprojet.tmol, someone might be able to work on it this holiday season (and this likely would not be part of a first pass). But in setup.py, you can set the cmdclass with a dict containing all the commands. But, better yet, `nuitka` seems to recommend just setting `build_with_nuitka=True` in your setup() call.

You can and still should specify a pyproject.toml like this:

```toml
[build-system]
requires = [
    "setuptools>=42",
    "wheel",
    "nuitka",
]
build-backend = "setuptools.build_meta"
```

**Reply by nschloe** [reply]
Thanks so much for looking into it! Adding `build_with_nuitka=True` indeed helped me getting this to work.
